allensdk/api/cloud_cache/cloud_cache.py

```diff
@@ -0,0 +1,305 @@
+
+        Returns
+        -------
+        bool
+            True if the file exists and is valid; False otherwise
+
+        Raises
+        -----
+        RuntimeError
+            If file_attributes.local_path exists but is not a file.
+            It would be unclear how the cache should proceed in this case.
+        """
+
+        if not os.path.exists(file_attributes.local_path):
+            return False
+        if not os.path.isfile(file_attributes.local_path):
+            raise RuntimeError(f"{file_attributes.local_path}\n"
+                               "exists, but is not a file;\n"
+                               "unsure how to proceed")
+
+        full_path = str(file_attributes.local_path.resolve())
+        test_checksum = file_hash_from_path(full_path)
+        if test_checksum != file_attributes.file_hash:
+            return False
+
+        return True
+
+    def _download_file(self, file_attributes: CacheFileAttributes) -> bool:
+        """
+        Check if a file exists and is in the expected state.
+
+        If it is, return True.
+
+        If it is not, download the file, creating the directory
+        where the file is to be stored if necessary.
+
+        If the download is successful, return True.
+
+        If the download fails (file hash does not match expectation),
+        return False.
+
+        Parameters
+        ----------
+        file_attributes: CacheFileAttributes
+            Describes the file to download
+
+        Returns
+        -------
+        boolean
+
+        Raises
+        ------
+        RuntimeError
+            If the path to the directory where the file is to be saved
+            points to something that is not a directory.
+        """
+
+        local_path = file_attributes.local_path
+        local_dir = safe_system_path(str(local_path.parents[0]))
+        if not os.path.exists(local_dir):
+            os.makedirs(local_dir)
+        if not os.path.isdir(local_dir):
+            raise RuntimeError(f"{local_dir}\n"
+                               "is not a directory")
+
+        bucket_name = bucket_name_from_uri(file_attributes.uri)
+        obj_key = relative_path_from_uri(file_attributes.uri)
+
+        n_iter = 0
+        max_iter = 10  # maximum number of times to try download
+
+        version_id = file_attributes.version_id
+
+        while not self._file_exists(file_attributes):
+            response = self.s3_client.get_object(Bucket=bucket_name,
+                                                 Key=str(obj_key),
+                                                 VersionId=version_id)
+
+            if 'Body' in response:
+                with open(local_path, 'wb') as out_file:
+                    for chunk in response['Body'].iter_chunks():
+                        out_file.write(chunk)
+
+            n_iter += 1
+            if n_iter > max_iter:
+                return False
+        return True
+
+    def data_path(self, file_id) -> pathlib.Path:
```

Comment:
should there be a `check_only` option?
the name of this function doesn't necessarily sound like it is going to do some downloading.

Reply:
Agree on that. Maybe `self._check_is_valid` instead of `self._download_file`

Reply:
Just to clarify:

Are you okay with the idea that `data_path(file_id, check_only=False)` downloads the file and returns the local path? I would like to prevent the user from having to handle the "does the file exist? 'No.' Then download it" logic.

Reply:
alternate to `check_only` would be `download_if_missing`
your idea above doesn't really get at the fact that `data_path` sounds like an attribute, when in fact is an action, and a potentially costly one.

```
for file_id in list_of_file_ids:
      print(data_path(file_id))
```
As a user, I'd want to know when I am downloading vs. other things. This seems to try to hide it.

Reply:
Naively, we could call this method `get_data` to reflect the "action vs. attribute" concern, but the method doesn't return the data. It just returns a path to where the data is stored (since "data" could mean any number of different file types)

Does the method name `access_data` communicate what is going on?

I still feel that the default behavior should be to download the data if it isn't there. The paths themselves are not going to be very interesting since they 

a) will contain the file_hash, making them difficult for humans to read
b) will point to nothing if your run with `check_only=True`/`download_if_missing=False`

Reply:
I'm starting to believe that `get_data` isn't horribly misleading, even if it only returns a path.

Is that sensible?

Reply:
perhaps I am just not grasping the use-case here. Maybe we should talk it over on a call to help me understand.

Reply:
There are now two methods

`data_path` returns a typed dict containing the local path to the file, a boolean indicating whether or not the file exists in a valid state, and the file's `CacheFileParameters` (this latter is included so that `download_data` can use `data_path` to get the info needed for download)

`download_data` actually downloads the file and returns its local path.